atomWithObservable can't recover from error (and also with any atoms?)

While working on #599, I noticed retrying from ErrorBoundary doesn't work (there's a failing test skipped). It's probably because the atom value (error) is stored in Provider, so re-render with the atom doesn't trigger a retry. If this is the case, we'd need to find a fundamental solution. 🤔 

Better to start with investigating how useState + useContext would work.

jotai v.1.2.1 apparently have this :

```
node_modules/jotai/utils/atomWithObservable.d.ts:13:5 - error TS1170: A computed property name in a type literal must refer to an expression whose type is a literal type or a 'unique symbol' type.

13     [Symbol.observable]?: () => ObservableLike<T>;
       ~~~~~~~~~~~~~~~~~~~

node_modules/jotai/utils/atomWithObservable.d.ts:13:13 - error TS2339: Property 'observable' does not exist on type 'SymbolConstructor'.

13     [Symbol.observable]?: () => ObservableLike<T>;
```

--- a/src/core/store.ts
+++ b/src/core/store.ts
@@ -177,6 +177,7 @@ export const createStore = (
       return
     }
     atomState.c?.() // cancel read promise
+    delete atomState.e // read error
     if (isInterruptablePromise(promise)) {
       atomState.p = promise // read promise
       delete atomState.c // this promise is from another atom state, shouldn't be canceled here

--- a/tests/error.test.tsx
+++ b/tests/error.test.tsx
@@ -492,7 +492,7 @@ describe('error recovery', () => {
     await findByText('Value: 1')
   })
 
-  it.only('recovers from async errors', async () => {
+  it('recovers from async errors', async () => {
     const asyncAtom = atom(async (get) => {
       const value = get(counter)
       await new Promise((resolve) => {

--- a/tests/error.test.tsx
+++ b/tests/error.test.tsx
@@ -26,7 +26,12 @@ class ErrorBoundary extends Component<
   }
   render() {
     return this.state.hasError ? (
-      <div>{this.props.message || 'errored'}</div>
+      <div>
+        {this.props.message || 'errored'}
+        <button onClick={() => this.setState({ hasError: false })}>
+          retry
+        </button>
+      </div>
     ) : (
       this.props.children
     )
@@ -445,5 +450,81 @@ describe('throws an error while updating in effect cleanup', () => {
 
     fireEvent.click(getByText('close'))
     expect(console.error).toHaveBeenCalledTimes(1)
+  })
+})
+
+describe('error recovery', () => {
+  const counter = atom(0)
+
+  const Counter = () => {
+    const [count, setCount] = useAtom(counter)
+    return <button onClick={() => setCount(count + 1)}>increment</button>
+  }
+
+  it('recovers from sync errors', async () => {
+    const syncAtom = atom((get) => {
+      const value = get(counter)
+
+      if (value === 0) {
+        throw new Error('An error occurred')
+      }
+
+      return value
+    })
+
+    const Display = () => {
+      return <div>Value: {useAtom(syncAtom)[0]}</div>
+    }
+
+    const { getByText, findByText } = render(
+      <Provider>
+        <Counter />
+        <ErrorBoundary>
+          <Display />
+        </ErrorBoundary>
+      </Provider>
+    )
+
+    await findByText('errored')
+
+    fireEvent.click(getByText('increment'))
+    fireEvent.click(getByText('retry'))
+    await findByText('Value: 1')
+  })
+
+  it.only('recovers from async errors', async () => {
+    const asyncAtom = atom(async (get) => {
+      const value = get(counter)
+      await new Promise((resolve) => {
+        setTimeout(resolve, 50)
+      })
+
+      if (value === 0) {
+        throw new Error('An error occurred')
+      }
+
+      return value
+    })
+
+    const Display = () => {
+      return <div>Value: {useAtom(asyncAtom)[0]}</div>
+    }
+
+    const { getByText, findByText } = render(
+      <Provider>
+        <Counter />
+        <ErrorBoundary>
+          <Suspense fallback={null}>
+            <Display />
+          </Suspense>
+        </ErrorBoundary>
+      </Provider>
+    )
+
+    await findByText('errored')
+
+    fireEvent.click(getByText('increment'))
+    fireEvent.click(getByText('retry'))
+    await findByText('Value: 1')
   })
 })

--- a/tests/error.test.tsx
+++ b/tests/error.test.tsx
@@ -454,16 +454,22 @@ describe('throws an error while updating in effect cleanup', () => {
 })
 
 describe('error recovery', () => {
-  const counter = atom(0)
+  const createCounter = () => {
+    const counterAtom = atom(0)
 
-  const Counter = () => {
-    const [count, setCount] = useAtom(counter)
-    return <button onClick={() => setCount(count + 1)}>increment</button>
+    const Counter = () => {
+      const [count, setCount] = useAtom(counterAtom)
+      return <button onClick={() => setCount(count + 1)}>increment</button>
+    }
+
+    return { Counter, counterAtom }
   }
 
   it('recovers from sync errors', async () => {
+    const { counterAtom, Counter } = createCounter()
+
     const syncAtom = atom((get) => {
-      const value = get(counter)
+      const value = get(counterAtom)
 
       if (value === 0) {
         throw new Error('An error occurred')
@@ -493,8 +499,9 @@ describe('error recovery', () => {
   })
 
   it('recovers from async errors', async () => {
+    const { counterAtom, Counter } = createCounter()
     const asyncAtom = atom(async (get) => {
-      const value = get(counter)
+      const value = get(counterAtom)
       await new Promise((resolve) => {
         setTimeout(resolve, 50)
       })

--- a/tests/utils/atomWithObservable.test.tsx
+++ b/tests/utils/atomWithObservable.test.tsx
@@ -1,4 +1,4 @@
-import { Component, Suspense } from 'react'
+import { Suspense } from 'react'
 import { fireEvent, render } from '@testing-library/react'
 import { Observable, Subject } from 'rxjs'
 import { useAtom } from '../../src/index'
@@ -6,31 +6,6 @@ import { atomWithObservable } from '../../src/utils'
 import { getTestProvider } from '../testUtils'
 
 const Provider = getTestProvider()
-
-class ErrorBoundary extends Component<
-  { message?: string },
-  { hasError: boolean }
-> {
-  constructor(props: { message?: string }) {
-    super(props)
-    this.state = { hasError: false }
-  }
-  static getDerivedStateFromError() {
-    return { hasError: true }
-  }
-  render() {
-    return this.state.hasError ? (
-      <div>
-        <div>{this.props.message || 'errored'}</div>
-        <button onClick={() => this.setState({ hasError: false })}>
-          retry
-        </button>
-      </div>
-    ) : (
-      this.props.children
-    )
-  }
-}
 
 it('count state', async () => {
   const observableAtom = atomWithObservable(
@@ -94,33 +69,4 @@ it('writable count state', async () => {
 
   fireEvent.click(getByText('button'))
   await findByText('count: 9')
-})
-
-// FIXME we would like to support retry
-it.skip('count state with error', async () => {
-  const myObservable = new Observable<number>((subscriber) => {
-    subscriber.error('err1')
-    subscriber.next(1)
-  })
-  const observableAtom = atomWithObservable(() => myObservable)
-
-  const Counter = () => {
-    const [state] = useAtom(observableAtom)
-
-    return <div>count: {state}</div>
-  }
-
-  const { findByText, getByText } = render(
-    <Provider>
-      <ErrorBoundary>
-        <Suspense fallback="loading">
-          <Counter />
-        </Suspense>
-      </ErrorBoundary>
-    </Provider>
-  )
-
-  await findByText('errored')
-  fireEvent.click(getByText('retry'))
-  await findByText('count: 1')
 })

